core/common/src/grpc_interceptor.rs

```diff
@@ -0,0 +1,238 @@
+    /// # Arguments
+    /// * `service_name` - The gRPC call's service name.
+    /// * `method_name` - The gRPC call's method name.
+    /// * `protobuf_message_bytes` - The request's protobuf messages as bytes.
+    fn handle_request(
+        &self,
+        service_name: &str,
+        method_name: &str,
+        protobuf_message: Bytes,
+    ) -> Result<Bytes, Box<dyn Error + Send + Sync>>;
+
+    /// Handle response. Return the new response.
+    ///
+    /// # Arguments
+    /// * `service_name` - The gRPC call's service name.
+    /// * `method_name` - The gRPC call's method name.
+    /// * `protobuf_message_bytes` - The response's protobuf messages as bytes.
+    fn handle_response(
+        &self,
+        service_name: &str,
+        method_name: &str,
+        protobuf_message: Bytes,
+    ) -> Result<Bytes, Box<dyn Error + Send + Sync>>;
+}
+
+/// This is the type that represents the factory method for gRPC Interceptors.
+type GrpcInterceptorFactory = fn() -> Box<dyn GrpcInterceptor + Send>;
+
+/// The tower layer that hosts a service that hosts a gRPC Interceptor.
+#[derive(Clone)]
+pub struct GrpcInterceptorLayer {
+    interceptor_factory: GrpcInterceptorFactory,
+}
+
+impl GrpcInterceptorLayer {
+    /// Create the tower layer for a gRPC Interceptor.
+    ///
+    /// # Arguments
+    /// * `interceptor_factory` - The factory method for creating the desired gRPC Interceptor.
+    pub fn new(interceptor_factory: GrpcInterceptorFactory) -> Self {
+        Self { interceptor_factory }
+    }
+}
+
+impl<S> Layer<S> for GrpcInterceptorLayer {
+    type Service = GrpcInterceptorService<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        GrpcInterceptorService { service, interceptor_factory: self.interceptor_factory }
+    }
+}
+
+/// The tower service that hosts a gRPC Interceptor.
+#[derive(Clone)]
+pub struct GrpcInterceptorService<S> {
+    service: S,
+    interceptor_factory: GrpcInterceptorFactory,
+}
+
+impl<S> GrpcInterceptorService<S> {
+    /// Retrieve the gRPC service name and method name from a URI.
+    ///
+    /// * `uri` - The uri used for the gRPC call.
+    fn retrieve_grpc_names_from_uri(uri: &Uri) -> (String, String) {
+        let uri_string = uri.to_string();
+        let uri_parts: Vec<&str> = uri_string.split('/').collect();
+        let mut service_name = String::new();
+        let mut method_name = String::new();
+        if uri_parts.len() == 5 {
+            method_name = uri_parts[4].to_string();
+            let qualified_service_name = uri_parts[3].to_string();
+            let qualified_service_name_parts: Vec<&str> =
+                qualified_service_name.split('.').collect();
+            if qualified_service_name_parts.len() == 2 {
+                service_name = qualified_service_name_parts[1].to_string();
+            }
+        }
+        (service_name, method_name)
+    }
+}
+
+impl<S> Service<http::request::Request<tonic::transport::Body>> for GrpcInterceptorService<S>
```

Comment:
I think it would be good to import the types used in this impl to help simplify it, this is a rather complex definition

Reply:
Import the types?

Reply:
yeah, like `use http::request::Request` and so on

Reply:
I can't do this without causing collisions